### PR TITLE
Feature/101 base reporter

### PR DESCRIPTION
**Pull request by lucamrgs**
Moved from other branch, supersedes PR #106

### CODE REVIEW

**Comment by lucamrgs** [comment]
HTTPOauth2 test failed due to 502 bad gateway :/

File: internal/reporter/reporter.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+// Reporter interfaces
+
+type IWorkflowReporter interface {
+	// -> Give info to downstream reporters
+	ReportWorkflow(executionContext execution.Metadata, playbook cacao.Playbook) error
+}
+type IStepReporter interface {
+	// -> Give info to downstream reporters
+	ReportStep(executionContext execution.Metadata, step cacao.Step, outVars cacao.Variables, err error) error
```
Comment (MaartendeKruijf) [inline internal/reporter/reporter.go:30]: remove the error return type

File: internal/reporter/reporter.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+func init() {
+	log = logger.Logger(component, logger.Info, "", logger.Json)
+}
+
+// Reporter interfaces
+
+type IWorkflowReporter interface {
+	// -> Give info to downstream reporters
+	ReportWorkflow(executionContext execution.Metadata, playbook cacao.Playbook) error
```
Comment (MaartendeKruijf) [inline internal/reporter/reporter.go:26]: Remove the error return type

File: test/unittest/decomposer/decomposer_test.go [hunk 3]
```diff
@@ -11,6 +11,7 @@ import (
 	"soarca/models/execution"
 	"soarca/test/unittest/mocks/mock_executor"
 	"soarca/test/unittest/mocks/mock_guid"
+	"soarca/test/unittest/mocks/mock_reporter"
```
Comment (MaartendeKruijf) [inline test/unittest/decomposer/decomposer_test.go:14]: Test that the reporter is called

File: test/unittest/executor/action/action_executor_test.go [hunk 4]
```diff
@@ -9,6 +9,7 @@ import (
 	"soarca/models/cacao"
 	"soarca/models/execution"
 	"soarca/test/unittest/mocks/mock_capability"
+	"soarca/test/unittest/mocks/mock_reporter"
```
Comment (MaartendeKruijf) [inline test/unittest/executor/action/action_executor_test.go:12]: Test if the reporter is called

File: test/unittest/mocks/mock_reporter/mock_downstream_reporter.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+	"github.com/stretchr/testify/mock"
+)
+
+type Mock_Downstream_Reporter struct {
+	mock.Mock
+}
+
+func (reporter *Mock_Downstream_Reporter) ReportWorkflow(workflowEntry downstream_reporter.WorkflowEntry) error {
+	return nil
```
Comment (MaartendeKruijf) [inline test/unittest/mocks/mock_reporter/mock_downstream_reporter.go:14]: Missing mack implementation like:

```
args := capability.Called(metadata, command, authentication, target, variables)
	return args.Get(0).(cacao.Variables), args.Error(1)
```

File: test/unittest/mocks/mock_reporter/mock_reporter.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+	"github.com/stretchr/testify/mock"
+)
+
+type Mock_Reporter struct {
+	mock.Mock
+}
+
+func (reporter *Mock_Reporter) ReportWorkflow(executionContext execution.Metadata, playbook cacao.Playbook) error {
+	return nil
```
Comment (MaartendeKruijf) [inline test/unittest/mocks/mock_reporter/mock_reporter.go:15]: Missing mock as well

**Review by MaartendeKruijf — CHANGES_REQUESTED** [review]

File: test/unittest/decomposer/decomposer_test.go (hunk 3 above)
Comment (MaartendeKruijf) [reply test/unittest/decomposer/decomposer_test.go:14]: Add to every test mock_reporter.assertExpectations

File: internal/reporter/reporter.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+func New(reporters []downstreamReporter.IDownStreamReporter) *Reporter {
+	instance := Reporter{}
+	if instance.reporters == nil {
+		instance.reporters = reporters
+	}
+	return &instance
+}
+
+func (reporter *Reporter) RegisterReporters(reporters []downstreamReporter.IDownStreamReporter) error {
+	maxReporters, _ := strconv.Atoi(utils.GetEnv("MAX_REPORTERS", strconv.Itoa(MaxReporters)))
```
Comment (MaartendeKruijf) [inline internal/reporter/reporter.go:51]: Maybe get this variable on NEW or just inject it?

**Review by MaartendeKruijf — CHANGES_REQUESTED** [review]
Please rebase and add mock assertions

File: internal/controller/controller.go [hunk 8]
```diff
@@ -67,10 +70,13 @@ func (controller *Controller) NewDecomposer() decomposer.IDecomposer {
 		}
 	}
 
-	actionExecutor := action.New(capabilities)
-	playbookActionExecutor := playbook_action.New(controller, controller)
+	// TODO: Instantiate reporters from config
```
Comment (MaartendeKruijf) [inline internal/controller/controller.go:73]: Remove comment and create issue

File: internal/decomposer/decomposer.go [hunk 9]
```diff
@@ -65,6 +68,13 @@ func (decomposer *Decomposer) Execute(playbook cacao.Playbook) (*ExecutionDetail
 	variables := cacao.NewVariables()
 	variables.Merge(playbook.PlaybookVariables)
 
+	// Reporting workflow instantiation
+	metadata := execution.Metadata{
```
Comment (MaartendeKruijf) [inline internal/decomposer/decomposer.go:72]: You're creating a metadata object here and only using the ExecutionId just pass that to ReportWorkflow

**Review by MaartendeKruijf — APPROVED** [review]

**Pull request merged**